### PR TITLE
[SLS] Filter out DD_TAGS, DD_EXTRA_TAGS & function specific tags from managed service spans

### PR DESCRIPTION
**Pull request by hghotra**
### What does this PR do?

This PR filters out the following tags from any managed service spans (also referred to as inferred spans):
- DD_TAGS (since they're meant to apply to the AWS Lambda & patched spans)
- DD_EXTRA_TAGS (since they're meant to apply to the AWS Lambda & patched spans)
- AWS function specific tags

Opening in place of https://github.com/DataDog/datadog-agent/pull/10420

### Motivation

It's confusing to have function specific tags in Managed Service spans (e.g. SQS, SNS, etc.)

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/agent/agent.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 	"time"
 
 	"github.com/DataDog/datadog-agent/pkg/obfuscate"
+	"github.com/DataDog/datadog-agent/pkg/serverless/trace/inferredspan"
```
Comment (maxday) [inline pkg/trace/agent/agent.go:15]: Up to now, we've managed to never import serverless package in non-serverless package.
I think you can hook to `modifySpan` as I think the only input you need is the span?

File: pkg/serverless/trace/inferredspan/inferred_span.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	// that lets us know whether this span should inherit its tags.
+	// Expected options are "lambda" and "self"
+	tagInferredSpanTagSource = "_inferred_span.tag_source"
+
+	// additional function keys to ignore
+	functionVersionTagKey = "function_version"
+	coldStartTagKey       = "cold_start"
+)
+
+var functionTagsToIgnore = [...]string{
```
Comment (maxday) [inline pkg/serverless/trace/inferredspan/inferred_span.go:23]: Is it possible to create an allow list instead?
I have the feeling that we will forget to add any new tags to this list and cause a breaking change by accident

File: pkg/trace/traceutil/span.go [hunk 3]
```diff
@@ -58,6 +58,16 @@ func SetMetric(s *pb.Span, key string, val float64) {
 	s.Metrics[key] = val
 }
 
+// SetMetaTags sets the tag map for a span
+func SetMetaTags(s *pb.Span, tags map[string]string) {
```
Comment (maxday) [inline pkg/trace/traceutil/span.go:62]: since Meta is already exposed, I think you can get rid of Set/Get?

File: pkg/serverless/trace/inferredspan/inferred_span.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+// _inferred_span.tag_source = "lambda" or missing => lambda related span
+func Check(span *pb.Span) bool {
+	if strings.Compare(span.Meta[tagInferredSpanTagSource], "self") == 0 {
+		return true
+	}
+	return false
+}
+
+// FilterFunctionTags filters out DD tags & function specific tags
+func FilterFunctionTags(input *map[string]string) {
```
Comment (maxday) [inline pkg/serverless/trace/inferredspan/inferred_span.go:48]: Maybe it's safer to return a new map[string]string, this way you're not modifying the input, so you're sure that no side effect could occur in this pure function

File: pkg/serverless/trace/inferredspan/inferred_span.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	tags.MemorySizeKey,
+	tags.ArchitectureKey,
+	functionVersionTagKey,
+	coldStartTagKey,
+}
+
+// Check determines if a span belongs to a managed service or not
+// _inferred_span.tag_source = "self" => managed service span
+// _inferred_span.tag_source = "lambda" or missing => lambda related span
+func Check(span *pb.Span) bool {
```
Comment (maxday) [inline pkg/serverless/trace/inferredspan/inferred_span.go:40]: nit : the body of this function can be replaced with `return strings.Compare(span.Meta[tagInferredSpanTagSource], "self") == 0`

File: pkg/serverless/trace/inferredspan/inferred_span.go (hunk 2 above)
Comment (hghotra) [reply pkg/serverless/trace/inferredspan/inferred_span.go:23]: No, unfortunately we don't know the full list of tags we should allow since the customer can potentially add custom tags. Forgetting to add any new tags to this list won't be a breaking change, it'll simply cause some unwanted tags to show up for managed service spans.. unless you meant breaking in another way.

File: pkg/serverless/tags/tags.go [hunk 6]
```diff
@@ -17,31 +17,48 @@ import (
 )
 
 const (
+	// Environment variables for Unified Service Tagging
```
Comment (nhinsch) [inline pkg/serverless/tags/tags.go:20]: 👍 Thanks for organizing these constants

File: pkg/serverless/tags/tags.go [hunk 7]
```diff
@@ -17,31 +17,48 @@ import (
...
+	FunctionARNKey = "function_arn"
+	// FunctionNameKey is the tag key for a function's name
+	FunctionNameKey = "functionname"
+	// ExecutedVersionKey is the tag key for a function's executed version
+	ExecutedVersionKey = "executedversion"
+	// RuntimeKey is the tag key for a function's runtime (e.g node, python)
+	RuntimeKey = "runtime"
+	// MemorySizeKey is the tag key for a function's allocated memory size
+	MemorySizeKey = "memorysize"
+	// ArchitectureKey is the tag key for a function's architecture (e.g. amd64, arm64)
```
Comment (nhinsch) [inline pkg/serverless/tags/tags.go:40]: Nit: I think AWS uses the tag x86_64 instead of amd64, might want to double check that though

File: pkg/serverless/trace/inferredspan/inferred_span.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	// that lets us know whether this span should inherit its tags.
+	// Expected options are "lambda" and "self"
+	tagInferredSpanTagSource = "_inferred_span.tag_source"
+
+	// additional function specific tag keys to ignore
+	functionVersionTagKey = "function_version"
+	coldStartTagKey       = "cold_start"
+)
+
+var functionTagsToIgnore = [...]string{
```
Comment (nhinsch) [inline pkg/serverless/trace/inferredspan/inferred_span.go:23]: I think including the `...` in between the `[]` before `string` means that this is now an array with length 11 instead of a slice.

Here is a blog post I found covering some of the differences: https://medium.com/@marty.stepien/arrays-vs-slices-bonanza-in-golang-fa8d32cd2b7c

tl;dr Slices are more flexible than arrays and I would always use a slice unless I had some special reason to use an array.

File: pkg/serverless/trace/inferredspan/inferred_span.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	tags.MemorySizeKey,
+	tags.ArchitectureKey,
+	functionVersionTagKey,
+	coldStartTagKey,
+}
+
+// Check determines if a span belongs to a managed service or not
+// _inferred_span.tag_source = "self" => managed service span
+// _inferred_span.tag_source = "lambda" or missing => lambda related span
+func Check(span *pb.Span) bool {
```
Comment (nhinsch) [inline pkg/serverless/trace/inferredspan/inferred_span.go:40]: I would rename this function so it's more clear what it's checking, and what `true` means. Something like `CheckIsInferredSpan` perhaps...

File: pkg/serverless/trace/inferredspan/inferred_span_test.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+func TestFilterFunctionTags(t *testing.T) {
+	tagsToFilter := map[string]string{"_inferred_span.tag_source": "self", "functionname": "lambda", "extra": "tag", "tag1": "value1"}
+
+	mockConfig := config.Mock()
+	mockConfig.Set("tags", []string{"tag1:value1"})
+	mockConfig.Set("extra_tags", []string{"extra:tag"})
+
+	assert.Equal(t, tagsToFilter["functionname"], "lambda")
+	assert.Equal(t, tagsToFilter["extra"], "tag")
+	assert.Equal(t, tagsToFilter["tag1"], "value1")
```
Comment (nhinsch) [inline pkg/serverless/trace/inferredspan/inferred_span_test.go:45]: What is the purpose of these three assertions?

File: pkg/serverless/trace/inferredspan/inferred_span_test.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+	filteredTags := FilterFunctionTags(tagsToFilter)
+
+	assert.Equal(t, filteredTags, map[string]string{"_inferred_span.tag_source": "self"})
+	assert.NotEqual(t, filteredTags["functionname"], "lambda")
+	assert.NotEqual(t, filteredTags["extra"], "tag")
+	assert.NotEqual(t, filteredTags["tag1"], "value1")
```
Comment (nhinsch) [inline pkg/serverless/trace/inferredspan/inferred_span_test.go:52]: Also, these three assertions seem redundant with the first one in this set?

File: pkg/serverless/trace/inferredspan/inferred_span_test.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+	checkLambdaSpanTagged := Check(lambdaSpanTagged)
+	checkLambdaSpanUntagged := Check(lambdaSpanUntagged)
+
+	assert.True(t, checkInferredSpan)
+	assert.False(t, checkLambdaSpanTagged)
+	assert.False(t, checkLambdaSpanUntagged)
+}
+
+func TestFilterFunctionTags(t *testing.T) {
+	tagsToFilter := map[string]string{"_inferred_span.tag_source": "self", "functionname": "lambda", "extra": "tag", "tag1": "value1"}
```
Comment (nhinsch) [inline pkg/serverless/trace/inferredspan/inferred_span_test.go:37]: I would add all of the tags we want to filter here to make sure we cover all of them. i.e. `architecture`, `memorysize`, etc. etc.

File: pkg/serverless/trace/inferredspan/inferred_span.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+// FilterFunctionTags filters out DD tags & function specific tags
+func FilterFunctionTags(input map[string]string) map[string]string {
+
+	if input == nil {
+		return nil
+	}
+
+	output := input
```
Comment (nhinsch) [inline pkg/serverless/trace/inferredspan/inferred_span.go:51]: It looks like this function returns an output map, but because maps are a reference type, I think you are also changing the input map here. To clarify this I would update this function so it doesn't return anything and just modifies the input.

File: pkg/serverless/trace/span_modifier.go [hunk 14]
```diff
@@ -19,4 +21,14 @@ func (s *spanModifier) ModifySpan(span *pb.Span) {
...
 		span.Service = s.tags["service"]
 	}
+	if inferredspan.Check(span) {
+		log.Debug("Detected a managed service span, filtering out function tags")
+
+		// filter out existing function tags inside span metadata
+		spanMetadataTags := span.Meta
+		if spanMetadataTags != nil {
+			inferredspan.FilterFunctionTags(spanMetadataTags)
+			span.Meta = spanMetadataTags
```
Comment (nhinsch) [inline pkg/serverless/trace/span_modifier.go:31]: Per my other comment, I think this isn't quite right:
- Get the metadata tags `span.Meta`
- Call `FilterFunctionTags` on `spanMetadataTags`. This returns something (which is ignored here) and *also* modifies the original `spanMetadataTags`
- Set span.Meta to `spanMetadatatags` -- but I don't think this actually doing anything, it was already set to that. The map has already been updated in place.

File: pkg/serverless/trace/inferredspan/inferred_span.go (hunk 13 above)
Comment (hghotra) [reply pkg/serverless/trace/inferredspan/inferred_span.go:51]: That's how I originally wrote this function but @maxday suggested we return a new object in favor of making this a pure function so I created an output map (but didn't deep copy it). I'll update the code so it deep copies the input map.

File: pkg/serverless/trace/span_modifier.go (hunk 14 above)
Comment (hghotra) [reply pkg/serverless/trace/span_modifier.go:31]: Yeah, this was working because of `output = input`, and originally the function being written in a way where `input` was directly modified.

**Review by nhinsch — APPROVED** [review]
